File: js/sessions.js
// sessions.js — Campaigns and sessions module.
const Sessions = {
  sessions: [],
  campaigns: [],
  characters: [],
  _lastLoaded: 0,
  _TTL: 3 * 60 * 1000,

  // Current modal state
  _editId: null,       // null = new, number = edit

  async loadCached() {
    if (Date.now() - this._lastLoaded < this._TTL) {
      this._render();
      return;
    }
    await this.load();
  },

  async load() {
    const status = document.getElementById('sessions-status');
    if (status) status.textContent = 'Ładowanie...';
    try {
      const [sessions, campaigns, chars] = await Promise.all([
        API.get('/sessions'),
        API.get('/campaigns'),
        API.get('/characters'),
      ]);
      this.sessions  = Array.isArray(sessions)  ? sessions  : [];
      this.campaigns = Array.isArray(campaigns) ? campaigns : [];
      this.characters = Array.isArray(chars)    ? chars     : [];
      this._lastLoaded = Date.now();
      if (status) status.textContent = '';
      this._render();
    } catch(e) {
      if (status) status.textContent = 'Błąd: ' + e.message;
    }
  },

  _invalidate() { this._lastLoaded = 0; },

  // ── Labels ────────────────────────────────────────────────────────────────
  _statusLabel(s) {
    return { recruiting: 'Nabór', ongoing: 'W trakcie', ended: 'Zakończona' }[s] || s;
  },
  _statusClass(s) {
    return { recruiting: 'sess-chip-recruiting', ongoing: 'sess-chip-ongoing', ended: 'sess-chip-ended' }[s] || '';
  },
  _riskLabel(r) {
    return { low: 'Niskie', moderate: 'Umiarkowane', high: 'Wysokie', extreme: 'Ekstremalne' }[r] || r;
  },
  _riskClass(r) {
    return { low: 'sess-risk-low', moderate: 'sess-risk-moderate', high: 'sess-risk-high', extreme: 'sess-risk-extreme' }[r] || '';
  },
  _scopeLabel(s) {
    return { intimate: 'Intymna', local: 'Lokalna', global: 'Globalna' }[s] || s;
  },

  // ── Render ────────────────────────────────────────────────────────────────
  _render() {
    const wrap = document.getElementById('sessions-list');
    if (!wrap) return;

    if (!this.sessions.length) {
      wrap.innerHTML = `<div class="tl-empty">
        <div style="font-size:2rem;margin-bottom:1rem;opacity:0.25;">⚔</div>
        Brak sesji. Kliknij <strong>＋ Nowa sesja</strong> aby dodać pierwszą.
      </div>`;
      return;
    }

    // Group by status order: recruiting → ongoing → ended
    const groups = [
      { key: 'recruiting', label: 'Nabór' },
      { key: 'ongoing',    label: 'W trakcie' },
      { key: 'ended',      label: 'Zakończone' },
    ];

    let html = '';
    groups.forEach(g => {
      const group = this.sessions.filter(s => s.status === g.key);
      if (!group.length) return;
      html += `<div class="sess-group">
        <div class="sess-group-label">${g.label} (${group.length})</div>
        ${group.map(s => this._renderCard(s)).join('')}
      </div>`;
    });

    wrap.innerHTML = html;
  },

  _renderCard(s) {
    const isOwner = State.currentUser &&
      (State.currentUser.role === 'admin' || State.currentUser.id === s.created_by);

    const charAvatars = (s.character_participants || []).map(p => {
      return p.avatar
        ? `<img class="tl-avatar tl-avatar-sm" src="${this._esc(p.avatar)}" title="${this._esc(p.name)}">`
        : `<div class="tl-avatar tl-avatar-sm tl-avatar-placeholder" title="${this._esc(p.name)}">${(p.name[0]||'?').toUpperCase()}</div>`;
    }).join('');

    const npcBadges = (s.npc_participants || []).map(n =>
      `<span class="sess-npc-badge">${this._esc(n)}</span>`
    ).join('');

    const dateStr = s.date_start
      ? (s.date_end ? `${s.date_start} → ${s.date_end}` : `od ${s.date_start}`)
      : '';

    const campaignBadge = s.campaign_name
      ? `<span class="sess-campaign-badge">📜 ${this._esc(s.campaign_name)}</span>` : '';

    return `<div class="sess-card">
      <div class="sess-card-header">
        <div class="sess-card-title">${this._esc(s.title)}</div>
        <div class="sess-card-chips">
          <span class="sess-chip ${this._statusClass(s.status)}">${this._statusLabel(s.status)}</span>
          <span class="sess-chip ${this._riskClass(s.risk)}">⚠ ${this._riskLabel(s.risk)}</span>
          <span class="sess-chip sess-chip-scope">◎ ${this._scopeLabel(s.scope)}</span>
        </div>
      </div>
      ${campaignBadge || dateStr ? `<div class="sess-card-meta">
        ${campaignBadge}
        ${dateStr ? `<span class="sess-date">📅 ${this._esc(dateStr)}</span>` : ''}
      </div>` : ''}
      ${s.description ? `<div class="sess-card-desc">${this._esc(s.description.slice(0, 200))}${s.description.length > 200 ? '…' : ''}</div>` : ''}
      ${charAvatars || npcBadges ? `<div class="sess-participants">
        <span class="tl-participants-label">Uczestnicy:</span>
        <div class="sess-participants-inner">${charAvatars}${npcBadges}</div>
      </div>` : ''}
      <div class="sess-card-footer">
        <span class="sess-creator">dodał/a ${this._esc(s.creator_display)}</span>
        ${isOwner ? `<div class="sess-card-actions">
          <button class="sm warn-btn" onclick="Sessions.openModal(${s.id})">✎ Edytuj</button>
          <button class="danger sm" onclick="Sessions.deleteSession(${s.id}, '${this._esc(s.title).replace(/'/g, "\\'")}')">✕</button>
        </div>` : ''}
      </div>
    </div>`;
  },

  // ── Modal ─────────────────────────────────────────────────────────────────
  openModal(sessionId = null) {
    this._editId = sessionId;

    const title = document.getElementById('sess-modal-title');
    if (title) title.textContent = sessionId ? 'Edytuj sesję' : 'Nowa sesja';

    // Populate campaign select
    this._renderCampaignSelect();

    // Populate character picker
    this._renderCharPicker([]);

    if (sessionId) {
      const s = this.sessions.find(x => x.id === sessionId);
      if (!s) return;
      document.getElementById('sess-f-title').value       = s.title;
      document.getElementById('sess-f-desc').value        = s.description || '';
      document.getElementById('sess-f-date-start').value  = s.date_start || '';
      document.getElementById('sess-f-date-end').value    = s.date_end   || '';
      document.getElementById('sess-f-new-campaign').value = '';

      // Set campaign select
      const campSel = document.getElementById('sess-f-campaign');
      if (campSel) campSel.value = s.campaign_id || '';

      // Set chips
      this._setChip('sess-status', s.status);
      this._setChip('sess-risk',   s.risk);
      this._setChip('sess-scope',  s.scope);

      // Set characters
      const charIds = (s.character_participants || []).map(p => p.character_id);
      this._renderCharPicker(charIds);

      // Set NPCs
      document.getElementById('sess-f-npcs').value =
        (s.npc_participants || []).join(', ');
    } else {
      // Defaults for new
      document.getElementById('sess-f-title').value       = '';
      document.getElementById('sess-f-desc').value        = '';
      document.getElementById('sess-f-date-start').value  = '';
      document.getElementById('sess-f-date-end').value    = '';
      document.getElementById('sess-f-new-campaign').value = '';
      document.getElementById('sess-f-npcs').value        = '';
      this._setChip('sess-status', 'recruiting');
      this._setChip('sess-risk',   'low');
      this._setChip('sess-scope',  'local');
    }

    document.getElementById('sess-modal-err').textContent = '';
    document.getElementById('sess-modal-overlay').classList.add('open');
  },

  closeModal() {
    document.getElementById('sess-modal-overlay')?.classList.remove('open');
  },

  _renderCampaignSelect() {
    const sel = document.getElementById('sess-f-campaign');
    if (!sel) return;
    sel.innerHTML = '<option value="">— bez kampanii —</option>' +
      this.campaigns.map(c =>
        `<option value="${c.id}">${this._esc(c.name)}${c.status === 'ended' ? ' (zakończona)' : ''}</option>`
      ).join('');
  },

  _renderCharPicker(selectedIds) {
    const wrap = document.getElementById('sess-char-picker');
    if (!wrap) return;
    wrap.innerHTML = this.characters.map(c => {
      const checked = selectedIds.includes(c.id);
      return `<label class="char-pick-row">
        <input type="checkbox" value="${c.id}" ${checked ? 'checked' : ''}>
        <span><strong>${this._esc(c.name)}</strong>
          <span style="color:var(--text-m);font-size:0.78rem;"> ${this._esc(c.owner_username || '')}</span>
        </span>
      </label>`;
    }).join('');
  },

  _setChip(groupId, value) {
    document.querySelectorAll(`[data-chip-group="${groupId}"]`).forEach(btn => {
      btn.classList.toggle('active', btn.dataset.chipValue === value);
    });
  },

  _getChip(groupId) {
    const active = document.querySelector(`[data-chip-group="${groupId}"].active`);
    return active?.dataset.chipValue || '';
  },

  toggleChip(groupId, value) {
    this._setChip(groupId, value);
  },

  toggleNewCampaign() {
    const sel = document.getElementById('sess-f-campaign');
    const wrap = document.getElementById('sess-new-campaign-wrap');
    if (!wrap) return;
    const show = sel?.value === '__new__';
    wrap.style.display = show ? 'block' : 'none';
  },

  async save() {
    const title = document.getElementById('sess-f-title')?.value.trim();
    if (!title) {
      document.getElementById('sess-modal-err').textContent = 'Podaj tytuł sesji.';
      return;
    }

    const campSel = document.getElementById('sess-f-campaign');
    const campaignId = (campSel?.value && campSel.value !== '__new__')
      ? parseInt(campSel.value) : null;
    const newCampaignName = document.getElementById('sess-f-new-campaign')?.value.trim() || '';

    const charIds = Array.from(
      document.querySelectorAll('#sess-char-picker input:checked')
    ).map(el => parseInt(el.value));

    const npcRaw = document.getElementById('sess-f-npcs')?.value || '';
    const npcNames = npcRaw.split(',').map(n => n.trim()).filter(Boolean);

    const payload = {
      title,
      description: document.getElementById('sess-f-desc')?.value.trim() || '',
      campaign_id: campaignId,
      new_campaign_name: newCampaignName,
      date_start: document.getElementById('sess-f-date-start')?.value || '',
      date_end:   document.getElementById('sess-f-date-end')?.value   || '',
      status: this._getChip('sess-status') || 'recruiting',
      risk:   this._getChip('sess-risk')   || 'low',
      scope:  this._getChip('sess-scope')  || 'local',
      character_ids: charIds,
      npc_names: npcNames,
    };

    const saveBtn = document.getElementById('sess-save-btn');
    if (saveBtn) { saveBtn.disabled = true; saveBtn.textContent = 'Zapisywanie...'; }

    try {
      let data;
      if (this._editId) {
        data = await API.put('/sessions/' + this._editId, payload);
      } else {
        data = await API.post('/sessions', payload);
      }
      if (data.error) throw new Error(data.error);
      this.closeModal();
      this._invalidate();
      await this.load();
    } catch(e) {
      document.getElementById('sess-modal-err').textContent = 'Błąd: ' + e.message;
    } finally {
      if (saveBtn) { saveBtn.disabled = false; saveBtn.textContent = 'Zapisz'; }
    }
  },

  async deleteSession(id, title) {
    if (!confirm(`Usunąć sesję „${title}"?`)) return;
    try {
      const data = await API.delete('/sessions/' + id);
      if (data.error) throw new Error(data.error);
      this._invalidate();
      await this.load();
    } catch(e) {
      alert('Błąd: ' + e.message);
    }
  },

  _esc(s) {
    return String(s || '').replace(/&/g,'&amp;').replace(/</g,'&lt;').replace(/>/g,'&gt;');
  },
};

File: routes/sessions.py
"""Sessions routes — campaigns and sessions CRUD."""
from flask import Blueprint, request, jsonify

from models import (
    db, Campaign, Session, SessionParticipant, Character,
    VALID_SESSION_STATUSES, VALID_SESSION_RISKS, VALID_SESSION_SCOPES,
)
from utils import get_current_owner, require_auth

sessions_bp = Blueprint("sessions", __name__)


def _can_edit_session(owner, session_obj):
    """Creator or admin can edit/delete a session."""
    return owner.role == "admin" or session_obj.created_by == owner.id


# ── Campaigns ─────────────────────────────────────────────────────────────────

@sessions_bp.route("/campaigns", methods=["GET"])
@require_auth
def list_campaigns():
    campaigns = Campaign.query.order_by(Campaign.name).all()
    return jsonify([c.to_dict() for c in campaigns])


@sessions_bp.route("/campaigns", methods=["POST"])
@require_auth
def create_campaign():
    owner = get_current_owner()
    body = request.get_json(silent=True) or {}
    name = str(body.get("name", "")).strip()[:200]
    if not name:
        return jsonify({"error": "Podaj nazwę kampanii"}), 400

    c = Campaign(
        name=name,
        description=str(body.get("description", "")).strip()[:2000] or None,
        status=body.get("status", "active") if body.get("status") in ("active", "ended") else "active",
        created_by=owner.id,
    )
    db.session.add(c)
    db.session.commit()
    return jsonify(c.to_dict()), 201


@sessions_bp.route("/campaigns/<int:cid>", methods=["PUT"])
@require_auth
def update_campaign(cid):
    owner = get_current_owner()
    c = db.session.get(Campaign, cid)
    if not c:
        return jsonify({"error": "Nie znaleziono kampanii"}), 404
    if owner.role != "admin" and c.created_by != owner.id:
        return jsonify({"error": "Brak uprawnień"}), 403

    body = request.get_json(silent=True) or {}
    if "name" in body:
        c.name = str(body["name"]).strip()[:200] or c.name
    if "description" in body:
        c.description = str(body["description"]).strip()[:2000] or None
    if body.get("status") in ("active", "ended"):
        c.status = body["status"]
    db.session.commit()
    return jsonify(c.to_dict())


@sessions_bp.route("/campaigns/<int:cid>", methods=["DELETE"])
@require_auth
def delete_campaign(cid):
    owner = get_current_owner()
    c = db.session.get(Campaign, cid)
    if not c:
        return jsonify({"error": "Nie znaleziono kampanii"}), 404
    if owner.role != "admin" and c.created_by != owner.id:
        return jsonify({"error": "Brak uprawnień"}), 403
    db.session.delete(c)
    db.session.commit()
    return jsonify({"ok": True})


# ── Sessions ──────────────────────────────────────────────────────────────────

@sessions_bp.route("/sessions", methods=["GET"])
@require_auth
def list_sessions():
    campaign_id = request.args.get("campaign_id", type=int)
    status = request.args.get("status", "").strip()

    q = Session.query
    if campaign_id:
        q = q.filter_by(campaign_id=campaign_id)
    if status in VALID_SESSION_STATUSES:
        q = q.filter_by(status=status)

    sessions = q.order_by(Session.date_start.desc().nullslast(), Session.id.desc()).all()
    return jsonify([s.to_dict() for s in sessions])


@sessions_bp.route("/sessions", methods=["POST"])
@require_auth
def create_session():
    owner = get_current_owner()
    body = request.get_json(silent=True) or {}

    title = str(body.get("title", "")).strip()[:200]
    if not title:
        return jsonify({"error": "Podaj tytuł sesji"}), 400

    # Optionally create a new campaign inline
    campaign_id = body.get("campaign_id")
    new_campaign_name = str(body.get("new_campaign_name", "")).strip()[:200]
    if new_campaign_name and not campaign_id:
        camp = Campaign(name=new_campaign_name, created_by=owner.id)
        db.session.add(camp)
        db.session.flush()
        campaign_id = camp.id

    if campaign_id and not db.session.get(Campaign, campaign_id):
        return jsonify({"error": "Nie znaleziono kampanii"}), 404

    status = body.get("status", "recruiting")
    risk   = body.get("risk",   "low")
    scope  = body.get("scope",  "local")
    if status not in VALID_SESSION_STATUSES: status = "recruiting"
    if risk   not in VALID_SESSION_RISKS:    risk   = "low"
    if scope  not in VALID_SESSION_SCOPES:   scope  = "local"

    s = Session(
        campaign_id = campaign_id,
        title       = title,
        description = str(body.get("description", "")).strip()[:5000] or None,
        date_start  = str(body.get("date_start", ""))[:10] or None,
        date_end    = str(body.get("date_end",   ""))[:10] or None,
        status      = status,
        risk        = risk,
        scope       = scope,
        created_by  = owner.id,
    )
    db.session.add(s)
    db.session.flush()

    _sync_participants(s, body.get("character_ids", []), body.get("npc_names", []))
    db.session.commit()
    return jsonify(s.to_dict()), 201


@sessions_bp.route("/sessions/<int:sid>", methods=["PUT"])
@require_auth
def update_session(sid):
    owner = get_current_owner()
    s = db.session.get(Session, sid)
    if not s:
        return jsonify({"error": "Nie znaleziono sesji"}), 404
    if not _can_edit_session(owner, s):
        return jsonify({"error": "Brak uprawnień"}), 403

    body = request.get_json(silent=True) or {}

    if "title" in body:
        s.title = str(body["title"]).strip()[:200] or s.title
    if "description" in body:
        s.description = str(body["description"]).strip()[:5000] or None
    if "date_start" in body:
        s.date_start = str(body["date_start"])[:10] or None
    if "date_end" in body:
        s.date_end = str(body["date_end"])[:10] or None
    if body.get("status") in VALID_SESSION_STATUSES:
        s.status = body["status"]
    if body.get("risk") in VALID_SESSION_RISKS:
        s.risk = body["risk"]
    if body.get("scope") in VALID_SESSION_SCOPES:
        s.scope = body["scope"]

    campaign_id = body.get("campaign_id")
    new_campaign_name = str(body.get("new_campaign_name", "")).strip()[:200]
    if new_campaign_name and not campaign_id:
        camp = Campaign(name=new_campaign_name, created_by=owner.id)
        db.session.add(camp)
        db.session.flush()
        s.campaign_id = camp.id
    elif "campaign_id" in body:
        s.campaign_id = campaign_id or None

    if "character_ids" in body or "npc_names" in body:
        _sync_participants(
            s,
            body.get("character_ids", [p.character_id for p in s.participants if p.character_id]),
            body.get("npc_names",     [p.npc_name     for p in s.participants if p.npc_name]),
        )

    db.session.commit()
    return jsonify(s.to_dict())


@sessions_bp.route("/sessions/<int:sid>", methods=["DELETE"])
@require_auth
def delete_session(sid):
    owner = get_current_owner()
    s = db.session.get(Session, sid)
    if not s:
        return jsonify({"error": "Nie znaleziono sesji"}), 404
    if not _can_edit_session(owner, s):
        return jsonify({"error": "Brak uprawnień"}), 403
    db.session.delete(s)
    db.session.commit()
    return jsonify({"ok": True})


@sessions_bp.route("/sessions/<int:sid>/participants", methods=["PUT"])
@require_auth
def update_participants(sid):
    owner = get_current_owner()
    s = db.session.get(Session, sid)
    if not s:
        return jsonify({"error": "Nie znaleziono sesji"}), 404
    if not _can_edit_session(owner, s):
        return jsonify({"error": "Brak uprawnień"}), 403

    body = request.get_json(silent=True) or {}
    _sync_participants(s, body.get("character_ids", []), body.get("npc_names", []))
    db.session.commit()
    return jsonify(s.to_dict())


# ── Helpers ───────────────────────────────────────────────────────────────────

def _sync_participants(session_obj, character_ids, npc_names):
    """Replace all participants with the given lists."""
    # Remove existing
    for p in list(session_obj.participants):
        db.session.delete(p)
    db.session.flush()

    # Add characters
    seen_chars = set()
    for cid in character_ids:
        try:
            cid = int(cid)
        except (TypeError, ValueError):
            continue
        if cid in seen_chars:
            continue
        if db.session.get(Character, cid):
            db.session.add(SessionParticipant(
                session_id=session_obj.id,
                character_id=cid,
            ))
            seen_chars.add(cid)

    # Add NPCs
    seen_npcs = set()
    for name in (npc_names or []):
        name = str(name).strip()[:200]
        if name and name not in seen_npcs:
            db.session.add(SessionParticipant(
                session_id=session_obj.id,
                npc_name=name,
            ))
            seen_npcs.add(name)
